File: aoc2019/aoc2019-java/src/test/java/de/havox_design/aoc2019/day16/Day16Test.java
package de.havox_design.aoc2019.day16;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

class Day16Test {

    @ParameterizedTest
    @MethodSource("getDataForTask1")
    void testTask1(String fileName, String expectation, int phases) {
        FlawedFrequencyTransmission objectUnderTest = new FlawedFrequencyTransmission(fileName);
        Assertions.assertEquals(expectation, objectUnderTest.processTask1(phases));
    }

    private static Stream<Arguments> getDataForTask1() {
        return Stream.of(
                Arguments.of("de/havox_design/aoc2019/day16/day16part1sample1.txt", "12345678", 0),
                Arguments.of("de/havox_design/aoc2019/day16/day16part1sample1.txt", "48226158", 1),
                Arguments.of("de/havox_design/aoc2019/day16/day16part1sample1.txt", "34040438", 2),
                Arguments.of("de/havox_design/aoc2019/day16/day16part1sample1.txt", "03415518", 3),
                Arguments.of("de/havox_design/aoc2019/day16/day16part1sample1.txt", "01029498", 4),
                Arguments.of("de/havox_design/aoc2019/day16/day16part1sample2.txt", "24176176", 100),
                Arguments.of("de/havox_design/aoc2019/day16/day16part1sample3.txt", "73745418", 100),
                Arguments.of("de/havox_design/aoc2019/day16/day16part1sample4.txt", "52432133", 100)
        );
    }

    @ParameterizedTest
    @MethodSource("getDataForTask2")
    void testTask2(String fileName, String expectation) {
        Assertions.assertEquals(expectation, FlawedFrequencyTransmission.processTask2(fileName));
    }

    private static Stream<Arguments> getDataForTask2() {
        return Stream.of(
                Arguments.of("de/havox_design/aoc2019/day16/day16part2sample1.txt", "84462026"),
                Arguments.of("de/havox_design/aoc2019/day16/day16part2sample2.txt", "78725270"),
                Arguments.of("de/havox_design/aoc2019/day16/day16part2sample3.txt", "53553731")
        );
    }
}

File: aoc2021/aoc2021-kotlin/src/test/kotlin/de/havox_design/aoc2021/day16/Day16Test.kt
package de.havox_design.aoc2021.day16

import de.havox_design.aoc.utils.kotlin.helpers.tests.shouldBe
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.math.BigInteger
import java.util.stream.Stream

class Day16Test {

    @ParameterizedTest
    @MethodSource("getDataForTestProcessPart1")
    fun testProcessPart1(filename: String, expectedResult: Int) =
        PacketDecoder(filename).processPart1().shouldBe(expectedResult)

    @ParameterizedTest
    @MethodSource("getDataForTestProcessPart2")
    fun testProcessPart2(filename: String, expectedResult: BigInteger) =
        PacketDecoder(filename).processPart2().shouldBe(expectedResult)

    companion object {
        @JvmStatic
        private fun getDataForTestProcessPart1(): Stream<Arguments> =
            Stream.of(
                Arguments.of("de/havox_design/aoc2021/day16/day16part1sample1.txt", 6),
                Arguments.of("de/havox_design/aoc2021/day16/day16part1sample2.txt", 9),
                Arguments.of("de/havox_design/aoc2021/day16/day16part1sample3.txt", 14),
                Arguments.of("de/havox_design/aoc2021/day16/day16part1sample4.txt", 16),
                Arguments.of("de/havox_design/aoc2021/day16/day16part1sample5.txt", 12),
                Arguments.of("de/havox_design/aoc2021/day16/day16part1sample6.txt", 23),
                Arguments.of("de/havox_design/aoc2021/day16/day16part1sample7.txt", 31)
            )

        @JvmStatic
        private fun getDataForTestProcessPart2(): Stream<Arguments> =
            Stream.of(
                Arguments.of("de/havox_design/aoc2021/day16/day16part2sample1.txt", BigInteger.valueOf(3)),
                Arguments.of("de/havox_design/aoc2021/day16/day16part2sample2.txt", BigInteger.valueOf(54)),
                Arguments.of("de/havox_design/aoc2021/day16/day16part2sample3.txt", BigInteger.valueOf(7)),
                Arguments.of("de/havox_design/aoc2021/day16/day16part2sample4.txt", BigInteger.valueOf(9)),
                Arguments.of("de/havox_design/aoc2021/day16/day16part2sample5.txt", BigInteger.valueOf(1)),
                Arguments.of("de/havox_design/aoc2021/day16/day16part2sample6.txt", BigInteger.valueOf(0)),
                Arguments.of("de/havox_design/aoc2021/day16/day16part2sample7.txt", BigInteger.valueOf(0)),
                Arguments.of("de/havox_design/aoc2021/day16/day16part2sample8.txt", BigInteger.valueOf(1))
            )
    }
}
